File: backend/api/failure.py
from fastapi import APIRouter, Depends, HTTPException

from backend.services.failure_service import FailureDemoService
from backend.dependencies import get_failure_service

router = APIRouter(
    prefix="/failure-demo",
    tags=["failure"],
)


@router.post("/razorpay")
async def ingest_razorpay_batch(
    failure_service: FailureDemoService = Depends(get_failure_service),
):
    """
    Fetch a Test Mode Razorpay batch and run RingWatch
    validation/quarantine handling.
    """
    try:
        return await failure_service.ingest_razorpay_batch()

    except RuntimeError as exc:
        raise HTTPException(
            status_code=502,
            detail=str(exc),
        ) from exc

    except Exception as exc:
        raise HTTPException(
            status_code=500,
            detail=f"Failure demo processing failed: {exc}",
        ) from exc


@router.post("/razorpay-synthetic")
async def generate_razorpay_synthetic_batch(
    failure_service: FailureDemoService = Depends(get_failure_service),
):
    """
    Generate a deterministic Razorpay-shaped test batch
    and run the same RingWatch validation/quarantine pipeline.
    """
    try:
        return await failure_service.ingest_synthetic_razorpay_batch()

    except RuntimeError as exc:
        raise HTTPException(
            status_code=500,
            detail=str(exc),
        ) from exc

    except Exception as exc:
        raise HTTPException(
            status_code=500,
            detail=f"Synthetic failure demo failed: {exc}",
        ) from exc

File: frontend/src/api/failure.js
import { apiFetch } from "./client";

/**
 * Fetch actual Test Mode payments from Razorpay
 * and run them through the RingWatch validation pipeline.
 */
export function ingestRazorpayBatch() {
  return apiFetch("/api/failure-demo/razorpay", {
    method: "POST",
  });
}

/**
 * Generate a 100-record Razorpay-shaped synthetic batch
 * with 10 deliberately malformed records.
 *
 * Expected:
 * 100 total
 * 90 valid
 * 10 malformed
 * 10 quarantined
 */
export function generateSyntheticFailureBatch() {
  return apiFetch("/api/failure-demo/razorpay-synthetic", {
    method: "POST",
  });
}